ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,253 @@
+            yield ClientSideOcrdFile(None, ID=f['file_id'], pageId=f['page_id'], fileGrp=f['file_grp'], url=f['url'], mimetype=f['mimetype'])
+
+    def find_all_files(self, *args, **kwargs):
+        return list(self.find_files(*args, **kwargs))
+
+    def add_agent(self, *args, **kwargs):
+        return self.session.request('POST', f'{self.url}/agent', data=OcrdAgentModel.create(**kwargs))
+
+    @property
+    def file_groups(self):
+        return self.session.request('GET', f'{self.url}/file_groups').json()['file_groups']
+
+    @deprecated_alias(pageId="page_id")
+    @deprecated_alias(ID="file_id")
+    def add_file(self, file_grp, content=None, file_id=None, url=None, mimetype=None, page_id=None, **kwargs):
+        r = self.session.request(
+            'POST',
+            self.url,
+            data=OcrdFileModel.create(
+                file_id=file_id,
+                file_grp=file_grp,
+                page_id=page_id,
+                mimetype=mimetype,
+                url=url).dict(),
+            files={'data': content}
+        )
+
+    def save(self):
+        self.session.request('PUT', self.url)
+
+
+#
+# Server
+#
+
+class OcrdMetsServer():
+
+    def __init__(self, workspace, host, port, socket):
+        self.workspace = workspace
+        self.host = host
+        self.port = port
+        self.socket = socket
+        self.log = getLogger('ocrd.workspace_client')
+
+
+    def startup(self):
+
+        # XXX HACK 
+        # circumventing dependency injection like this is bad and
+        # needs to be refactored once it's all runnign
+        workspace = self.workspace
+
+        app = FastAPI(
+            title="OCR-D METS Server",
+            description="Providing simultaneous write-access to mets.xml for OCR-D",
+        )
+
+        @app.exception_handler(ValidationError)
+        async def exception_handler_invalid400(request: Request, exc: ValidationError):
+            return JSONResponse(status_code=400, content=exc.errors())
+
+        @app.exception_handler(FileExistsError)
+        async def exception_handler_invalid400(request: Request, exc: FileExistsError):
+            return JSONResponse(status_code=400, content=str(exc))
+
+        @app.exception_handler(re.error)
+        async def exception_handler_invalid400(request: Request, exc: re.error):
+            return JSONResponse(status_code=400, content=f'invalid regex: {exc}')
+
+        @app.get("/", response_model=OcrdFileListModel)
+        async def find_files(
+            file_grp : Union[str, None] = None,
+            file_id : Union[str, None] = None,
+            page_id : Union[str, None] = None,
+            mimetype : Union[str, None] = None,
+        ):
+            """
+            Find files in the mets
+            """
+            found = workspace.mets.find_all_files(fileGrp=file_grp, ID=file_id, pageId=page_id, mimetype=mimetype)
+            return OcrdFileListModel.create(found)
+
+        @app.put('/')
```

Comment:
I like the save with the put to `/` though, even if it doesn't completely fit to what I said to the other requests to `/`